sdk/queue/priority_queue.go

```diff
@@ -0,0 +1,206 @@
+		Key: key,
+	}
+}
+
+// New initializes the internal data structures and returns a new
+// PriorityQueue
+func New() *PriorityQueue {
+	pq := PriorityQueue{
+		data:    make(pQueue, 0),
+		dataMap: make(map[string]*Item),
+	}
+	heap.Init(&pq.data)
+	return &pq
+}
+
+// PriorityQueue facilitates queue of Items, providing Push, Pop, and
+// PopByKey convenience methods. The ordering (priority) is an int64 value
+// with the smallest value is the highest priority. PriorityQueue maintains both
+// an internal slice for the queue as well as a map of the same items with their
+// keys as the index. This enables users to find specific items by key. The map
+// must be kept in sync with the data slice.
+// See https://golang.org/pkg/container/heap/#example__priorityQueue
+type PriorityQueue struct {
+	// data is the internal structure that holds the queue, and is operated on by
+	// heap functions
+	data pQueue
+
+	// dataMap represents all the items in the queue, with unique indexes, used
+	// for finding specific items. dataMap is kept in sync with the data slice
+	dataMap map[string]*Item
+
+	// mapMutex is used to facilitate read/write locks on the dataMap
+	dataMutex sync.Mutex
+}
+
+// pQueue is the internal data structure used to satisfy heap.Interface. This
+// prevents users from calling Pop and Push heap methods directly
+type pQueue []*Item
+
+var _ heap.Interface = &pQueue{}
+var _ sort.Interface = &pQueue{}
+
+// Item is something managed in the priority queue
+type Item struct {
+	// Key is a unique string used to identify items in the internal data map
+	Key string
+	// Value is an unspecified type that implementations can use to store
+	// information
+	Value interface{}
+
+	// Priority determines ordering in the queue, with the lowest value being the
+	// highest priority
+	Priority int64
+
+	// index is an internal value used by the heap package, and should not be
+	// modified by any consumer of the priority queue
+	index int
+}
+
+// Pop pops the highest priority item from the queue. This is a
+// wrapper/convenience method that calls heap.Pop, so consumers do not need to
+// invoke heap functions directly
+func (pq *PriorityQueue) Pop() (*Item, error) {
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if pq.Len() == 0 {
+		return nil, ErrEmpty
+	}
+	item := heap.Pop(&pq.data).(*Item)
+	delete(pq.dataMap, item.Key)
+	return item, nil
+}
+
+// Push pushes an item on to the queue. This is a wrapper/convenience
+// method that calls heap.Push, so consumers do not need to invoke heap
+// functions directly. Items must have unique Keys, and Items in the queue
+// cannot be updated. To modify an Item, users must first remove it and re-push
+// it after modifications
+func (pq *PriorityQueue) Push(i *Item) error {
+	if i.Key == "" {
+		return errors.New("error adding item: Item Key is required")
+	}
+
+	pq.dataMutex.Lock()
+	defer pq.dataMutex.Unlock()
+
+	if _, ok := pq.dataMap[i.Key]; ok {
+		return ErrDuplicateItem
```

Comment:
I wonder if it would be better to replace the value instead of error? What do you think?

Reply:
Done in 41ffc88d1

Reply:
It's possible that an Item in the queue has a `WAL ID` being tracked in it's `Value` field, in which case replacing the Item after, for example a user updates the `rotation_period`, that value would be lost 🤔 

It seems like it could be very edge-case though? Explicit pop'ing seemed "safer" to me, but I'm willing to consider replacing if you have more thoughts here.

Reply:
Sure, that sounds fine as is then.